design/ceph-deprecate-old-osd-types.md

```diff
@@ -0,0 +1,212 @@
+directories like `/run/udev`.
+
+Similar to the current Rook OSD prepare binary, this proposal calls for a Rook inspection binary
+which will examine the disk hardware, rejecting disks that are already in use by applications other
+than Rook (as per current behavior). The inspection binary stores a host device report in a
+ConfigMap which will be read by the operator. The report need merely contain a list of devices which
+are able to be provisioned for Ceph's use (e.g., `/dev/sdc` or `/dev/disk/by-id/blah`).
+
+Because `ceph-volume` is ultimately responsible for preparing the devices, it is ideal if
+`ceph-volume` does this inspection itself and produces a report detailing the disks it could use
+for later provisioning. The inspection binary should store `ceph-volume`'s report into the
+aforementioned ConfigMap.
+
+As a matter of future security, the pod responsible for inspection should have read-only access to
+`/dev` (and `/run/udev` if needed). This stage should not have write access to any part of the host
+unless it is absolutely required.
+
+#### Stage 2: Process device inspection report
+In the second stage, the operator examine the host device inspection report in the ConfigMap. The
+primary inspection processing done would be to filter the list of devices reported based on the
+`deviceFilter` cluster storage setting. This stage can also compare the host `devices` specified in
+the cluster config to the devices found during inspection.
+
+The insertion of stage 2 executed within the operator itself, placed in between stages 1 and 3
+allows for easy security improvements later, and it also helps keep the majority of the Ceph
+configuration logic in the operator itself rather than distributing the logic between multiple
+binaries. Existing logic distribution has contributed to operation which is not always transparent
+or easy to extend.
+
+#### Stage 3: Host device preparation
+This stage is one in which `ceph-volume` provisions the devices. As with stage 1, this design calls
+for a Rook OSD preparation binary. The binary's single purpose is to call `ceph-volume` to provision
+devices and store a post-preparation device report from `ceph-volume` into a ConfigMap (possibly the
+same ConfigMap from step 1) for the operator to inspect later.
+
+The binary should act as a simple pass-through for `ceph-volume` arguments to keep the binary's
+functions as small as possible; the Ceph operator should contain the majority of the logic related
+to Ceph's configuration. Arguments to `ceph-volume` here will be based on the host devices found in
+stage 1 and processing done in stage 2.
+
+For future security, this preparation pod should only need access to the individual disks which
+`ceph-volume` will use to provision Ceph.
+
+#### Stage 4: Start OSDs
+The last step remaining is to start the OSDs previously created by `ceph-volume`. The
+post-preparation device report stored in step 3 will be used to determine which osds to start on any
```

Comment:
I wonder if instead of a post-preparation report, it makes sense to skip back to step 1.  Inspection should then report prepared+ready devices that are just missing the OSD pods.  This would avoid some of the complexity around the post-preparation report and ceph-volume incantations that, say, produce 4 OSDs for a single device.

Reply:
That makes a lot of sense. Then we can call `ceph-volume` directly from a pod entrypoint and reduce the number of supplementary binaries Rook has. I'll include this in my next go-round.